open-amp: resource table: Add support for vendor-specific features

iuliana-prodan opened the pull request:
This update introduces support for vendor-specific features within the OpenAMP remote processor resource table. 

To facilitate the integration of custom data structures, new macros have been added to define initialization routines for resource table entries.

The core resource table has been refactored to include hooks that enable vendor-specific initialization logic. As part of this enhancement, NXP-specific support has been implemented, incorporating custom definitions and initialization routines tailored to NXP’s communication requirements.

To demonstrate this capability, the imx8ulp_evk_mimx8ud7_adsp board configuration has been updated to activate vendor-specific resource table features.

This enhancement allows better flexibility for remote processors requiring custom features configurations.

uLipe commented on lib/open-amp/Kconfig:
```diff
@@ -34,3 +34,35 @@ config OPENAMP_RSC_TABLE_IPM_TX_ID
+config OPENAMP_RSC_TABLE_VENDOR_RSC_LENGTH
+	int "Vendor specific length of the resource"
+	default 0
+	range 0 4
```
Question instead of using the size to actually reserve space in the rsc_table, why not create a specific option let's ADD_VENDOR_SPEFICIC_AREA?

I'm afraid that implementors may use for example 0 - sized values and it silently have hard time to debug when this value is badly populated

Having a switch like option enforces the implementor to search the documentation to refer the "right" way to populate the vendor reserved area.

iuliana-prodan replied:
Instead of CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_0, which means the vendor-specific feature is disable, I can add:
```
config CONFIG_OPENAMP_RSC_TABLE_VENDOR_RSC_FEATURE 
	bool "Vendor specific resource entry"
	depends on OPENAMP_RSC_TABLE
	help
	  Enable vendor-specific features supported by remote processor.
```
Will that be ok?

uLipe replied:
Hi @iuliana-prodan , I like that approach and it sounds more explicit on what this symbol does :)

uLipe reviewed (commented):
I have a question regarding turning this option on and off, otherwise looks good to me.

arnopo commented on lib/open-amp/resource_table.c:
```diff
@@ -72,6 +76,26 @@ static struct fw_resource_table __resource resource_table = {
+			, CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_3,
+#endif
+	},
+#endif
```
Regarding the implementation I wonder if you should not define your own table in you SoC Adapting this for all vendors could become a nightmare.

This one could be the default one without specific vendor resources 

An example of implementation would be to define a `OPENAMP_VENDOR_ RSC_TABLE` config that would include a `vendor_resource_table.h` instead of defining the `enum rsc_table_entries` and ` struct fw_resource_table  `

Some other/better probably exist...

iuliana-prodan replied:
@carlocaione Can you please review since you are the assignee & maintainer?
Thank you!

iuliana-prodan commented:
Changes in v2:
- added `CONFIG_OPENAMP_RSC_TABLE_VENDOR_RSC_FEATURE`, instead of `CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_0` for enablement;
- added `CONFIG_RSC_VENDOR_TYPE` to specify vendor specific type number, range 128-512.

iuliana-prodan commented:
An example on how this will be used is the i.MX8ULP support for `openamp_rsc_table` sample.
We just need to set some configs in the dts overlay of the sample, for this particular board:
```
diff --git a/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf b/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
index 88f6a2df451..decb9495d52 100644
--- a/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
+++ b/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
@@ -9,3 +9,13 @@ CONFIG_MBOX_NXP_IMX_MU=y
 
 CONFIG_OPENAMP_ADDR_TRANSLATION=y
 CONFIG_SHELL=n
+
+CONFIG_OPENAMP_RSC_TABLE_VENDOR_RSC_FEATURE=y
+CONFIG_RSC_VENDOR_TYPE=128
+CONFIG_OPENAMP_RSC_TABLE_VENDOR_RSC_LENGTH=3
+#magic number
+CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_0=0x6E787073
+#version
+CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_1=0x0
+#supported features - request host NOT to wait for FW_READY reply
+CONFIG_OPENAMP_RSC_TABLE_VENDOR_DATA_FEATURES_2=0x1
```
I haven't created a PR yet, for this change since there are still some dependencies that are not merged.

iuliana-prodan commented:
Changes in `v3`:
-  add a new header file to define initialization for OpenAMP resource table;
- add vendor-specific config options;
- update build system to include vendor-specific files, if defined;
- add NXP-specific resource table implementation.

An example on how to use this, and how to enable it is:
```
diff --git a/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf b/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
index 88f6a2df451..ae5017e56f9 100644
--- a/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
+++ b/samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.conf
@@ -9,3 +9,7 @@ CONFIG_MBOX_NXP_IMX_MU=y
 
 CONFIG_SHELL=n
+
+CONFIG_OPENAMP_VENDOR_RSC_TABLE_ENTRY=y
+CONFIG_OPENAMP_VENDOR_RSC_TYPE=128
+CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE="nxp_resource_table.c"
```
This diff will be appended to this PR, when PR https://github.com/zephyrproject-rtos/zephyr/pull/89317 will be merged.

arnopo commented on lib/open-amp/rsc_table_init.h:
```diff
@@ -0,0 +1,60 @@
+	.vendor_type = CONFIG_OPENAMP_VENDOR_RSC_TYPE,
+#endif
+
+
```
extra line to remove

arnopo commented on lib/open-amp/rsc_table_init.h:
```diff
@@ -0,0 +1,60 @@
+#ifndef RESOURCE_TABLE_INIT_H__
+#define RESOURCE_TABLE_INIT_H__
+
+	.hdr = {
```
should be a Macro

```
# define RESOURCE_TABLE_INIT  \
{ \
	 .hdr = {  \
....
}	 
```

arnopo commented on lib/open-amp/CMakeLists.txt:
```diff
@@ -5,4 +5,10 @@
+  zephyr_sources(vendor_rsc_table/${CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE})
+elseif(CONFIG_OPENAMP_RSC_TABLE)
+  zephyr_sources(resource_table.c)
+endif()
```
what about having

```
if(CONFIG_OPENAMP_VENDOR_RSC_TABLE)
# include vendor-specific resource table files
  zephyr_sources(${CONFIG_OPENAMP_VENDOR_RSC_TABLE_DIR/resource_table.c})
  zephyr_include_directories(${CONFIG_OPENAMP_VENDOR_RSC_TABLE_DIR})
elseif(CONFIG_OPENAMP_RSC_TABLE)
# include default resource table files
  zephyr_include_directories_ifdef(CONFIG_OPENAMP_RSC_TABLE .)
  zephyr_sources(resource_table.c)
endif()
```
CONFIG_OPENAMP_VENDOR_RSC_TABLE_DIR define the vendor directory  where the resource table is defined

Then you can select your own resource table an let resource_table.c unchanged

arnopo commented on lib/open-amp/vendor_rsc_table/nxp_resource_table.c:
```diff
@@ -0,0 +1,49 @@
+		.magic_num = 0x6E787073,
+		.version = 0x0,
+		.features = 0x1,
+	},
```
Seems that you need more that a vendor specific type, so you have a specific resource table. 
redefining the `fw_resource_table` seems making sense instead of defining a nxp_resource_table type

iuliana-prodan replied:
Redefining the whole `fw_resource_table` is a lot of code duplication. Also, if something changes in the base resource table the vendor specific needs to be updated, it will be hard to maintain.
All that I add in the resource table is just a new vendor specific entry. Other vendors can come with their own entry on top of the base resource table from `rsc_table_init.h`

iuliana-prodan commented on lib/open-amp/resource_table.c:
```diff
@@ -40,44 +40,7 @@ extern char ram_console_buf[];
-		"Zephyr_log",
-	},
-#endif
+	#include "rsc_table_init.h"
```
I've tried this, I cannot use macro since in the initialization of resource table we have a few defines, like:
```
...
	.offset = {

#if (CONFIG_OPENAMP_RSC_TABLE_NUM_RPMSG_BUFF > 0)
		offsetof(struct fw_resource_table, vdev),
#endif

#if defined(CONFIG_RAM_CONSOLE)
		offsetof(struct fw_resource_table, cm_trace),
#endif
...
```

iuliana-prodan commented:
Hi @arnopo,
I've updated the PR to use macros as you suggested.
Could you please take a look when you have a moment and let me know if any further changes are needed?
Thanks!

arnopo reviewed (commented):
Still few concerns around the flexibility/maintanability of the  solution.
I proposed an alternative, don't hesitate to ping me by discord if it isnot clear or if you see some blockers

iuliana-prodan committed:
lib: open-amp: Replace struct fw_resource_table with void

Switching to void allows greater flexibility in supporting
vendor-specific resource tables.

This change reverts commit 39863b66bdfb, and adds some fixes.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
lib: open-amp: Allow vendor specific resource table definition

Enhance flexibility by allowing vendors to define their own
resource tables. This is achieved by moving the vdev and vring
functions into the source file, requiring each vendor to
implement these functions within their specific resource
table definitions.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
lib: open-amp: Add base support for resource table

Introduce macros to define initialization
routines for OpenAMP resource table entries.

This will allow to extend the default OpenAMP
resource table with vendor-specific data structures.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
Changes in `v5`:

- Updated macros to use capital letters for better readability, following @arnopo’s suggestion.
- Introduced a vendor-specific resource table instead of just a vendor-specific entry. This allows for adding multiple vendor entries in any order within the resource table.
- Removed `OPENAMP_VENDOR_RSC_TYPE`; each vendor will now define this in their own resource table.
- To support vendor-specific resource tables, replaced the `struct fw_resource_table` with a `void` pointer for the resource table reference.

arnopo commented on lib/open-amp/vendor_rsc_table/nxp_resource_table.c:
```diff
@@ -0,0 +1,74 @@
+/*
```
could be nice to use similar tree than dts folder  by naming the folder `vendor` instead of `vendor_rsc_table`

arnopo commented on lib/open-amp/CMakeLists.txt:
```diff
@@ -5,4 +5,11 @@
+elseif(CONFIG_OPENAMP_RSC_TABLE)
+# include generic resource table
+  zephyr_sources(resource_table.c)
+endif()
```
seems that `CONFIG_OPENAMP_VENDOR_RSC_TABLE_ENTRY` is now uselless, or i missed something? 

```
if(CONFIG_OPENAMP_RSC_TABLE)
if(CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE)
  zephyr_include_directories(vendor_rsc_table)
  zephyr_sources(vendor_rsc_table/${CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE})
else()
  # include generic resource table
  zephyr_sources(resource_table.c)
endif()
endif()
```

iuliana-prodan replied:
I've updated the code and now `CONFIG_OPENAMP_VENDOR_RSC_TABLE` depens on `CONFIG_OPENAMP_RSC_TABLE`.

arnopo reviewed (commented):
This approach seems  a good compromise. No more design issue on my side.
Only 2 minor comments

iuliana-prodan committed:
lib: open-amp: Add NXP-specific resource table implementation

Implement NXP-specific support for the remote processor
resource table, including custom definitions and initialization
logic to meet NXP's communication requirements.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
lib: open-amp: Add vendor-specific config options

Add config options for vendor-specific resource tables.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
lib: open-amp: Update CMakeLists.txt for vendor resource table

Modify the build system to conditionally compile
vendor-specific resource table support based on configuration.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

arnopo commented on lib/open-amp/Kconfig:
```diff
@@ -47,3 +61,11 @@ config OPENAMP_COPY_RSC_TABLE
+	depends on OPENAMP_VENDOR_RSC_TABLE
+	help
+	  Name of a source file containing vendor-specific
+	  resource table.
```
Look to me that there are some redundancies in configs

Could we do it in this way:

`OPENAMP_RSC_TABLE`  should specify if there is a resource table
`OPENAMP_VENDOR_RSC_TABLE_FILE` point to the resource table file that is the legacy file by default


```
config OPENAMP_VENDOR_RSC_TABLE_FILE
	string "Source file containing vendor-specific resource table"
	default ""
	depends on OPENAMP_RSC_TABLE
	help
	  Name of a source file containing vendor-specific  resource table.
```

and then in CMakeLists.txt
zephyr_include_directories_ifdef(CONFIG_OPENAMP_RSC_TABLE_PRESENT .)
# include vendor-specific resource table files
if(CONFIG_OPENAMP_VENDOR_RSC_TABLE STREQUAL "")
# include generic resource table
  zephyr_sources(resource_table.c)
ielseif(CONFIG_OPENAMP_VENDOR_RSC_TABLE )
  zephyr_include_directories(vendor)
  zephyr_sources(vendor/${CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE})
endif()

iuliana-prodan replied:
>Look to me that there are some redundancies in configs

Here are the reasons why I wrote the code as I did:

Using `CONFIG_OPENAMP_RSC_TABLE_NUM_RPMSG_BUFF` in vendor rsc table is not right. 
For the generic table this config enables the VDEV_ENTRY that is mandatory for the sample.

I've made the `CONFIG_OPENAMP_RSC_TABLE` and `CONFIG_OPENAMP_VENDOR_RSC_TABLE` exclusive so a vendor can use only what is needed in their resource table. 

The only configs that are needed in both generic and vendor rsc table will have the `CONFIG_OPENAMP_RSC_TABLE_PRESENT`.

>if(CONFIG_OPENAMP_VENDOR_RSC_TABLE STREQUAL "")

In C, you can't directly check whether a Kconfig string is empty or not to conditionally include code. This limitation affects what I would need to implement [here ](https://github.com/zephyrproject-rtos/zephyr/pull/87296/files#diff-a85d23a295b1fb7f99cdcbddebde04b826470ec5d5d1c45ab712338680bff103R18) or [here](https://github.com/zephyrproject-rtos/zephyr/pull/87296/files#diff-a85d23a295b1fb7f99cdcbddebde04b826470ec5d5d1c45ab712338680bff103R112) if I go with your proposed solution.

iuliana-prodan committed:
samples: Enable vendor-specific support for i.MX8ULP

Update the imx8ulp_evk_mimx8ud7_adsp board configuration to enable
vendor-specific resource table features.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
Changes in `v7`:
- made `CONFIG_OPENAMP_VENDOR_RSC_TABLE` dependent on `CONFIG_OPENAMP_RSC_TABLE`;
- removed the patch that modified `linked.ld` for 8ULP, as the dependency is now handled via Kconfig.

In this PR I've intreduced only 2 new configuration options:
`CONFIG_OPENAMP_VENDOR_RSC_TABLE`: enables vendor-specific resource table support.
`CONFIG_OPENAMP_VENDOR_RSC_TABLE_FILE`: specifies the file implementing the vendor-specific resource table.

arnopo reviewed (commented):
a last comment, with that LGTM

iuliana-prodan commented:
Changes in `v8`:
- move `zephyr_include_directories` for vendor headers under condition `CONFIG_OPENAMP_VENDOR_RSC_TABLE`.
